portiaui/app/models/base.js

```diff
@@ -3,8 +3,110 @@ import DS from 'ember-data';
 const { get } = Ember;
 
+let currentActionModel = null;
+const modelActionQueue = [];
+
+function runActions() {
+    if (currentActionModel && currentActionModel.get('isSaving')) {
+        // current action has not finished
+        return;
+    }
+    currentActionModel = null;
+    while (modelActionQueue.length) {
+        const nextModel = modelActionQueue.shift();
+        if (nextModel.get('isSaving')) {
+            // model has started saving again, wait for the next didCommit event
+            return;
+        }
+        if (nextModel.get('isDeleted') && !nextModel.get('hasDirtyAttributes')) {
+            // model has been deleted
+            continue;
+        }
+        if (nextModel.get('pendingDelete')) {
+            nextModel.deleteRecord();
+            nextModel.set('pendingDelete', false);
+        }
+        const pendingSave = nextModel.get('pendingSave');
+        if (pendingSave) {
+            nextModel.set('pendingSave', null);
+            const {resolver, options} = pendingSave;
+            // apply the save and resolve the promise, then wait for the next didCommit event
+            nextModel.save(options).then(resolver.resolve, resolver.reject);
+            return;
+        }
+    }
+}
+
+function mergeSaveOptions(dst, src) {
+    if (!dst) {
+        return src;
+    }
+
+    if (!src) {
+        src = {};
+    }
+
+    if (src.coalesce) {
+        if (dst.coalesce) {
+            for (let coalesce of src.coalesce) {
+                const {model, options} = coalesce;
+                const matchingCoalesce = dst.coalesce.findBy('model', model);
+                if (matchingCoalesce) {
+                    matchingCoalesce.options = mergeSaveOptions(matchingCoalesce.options, options);
+                } else {
+                    dst.coalesce.push(coalesce);
+                }
+            }
+        } else {
+            dst.coalesce = src.coalesce;
+        }
+    }
+
+    if (src.partial) {
+        if (dst.partial) {
+            for (let field of src.partial) {
+                if (!dst.partial.includes(field)) {
+                    dst.partial.push(field);
+                }
+            }
+        }
+        // else dst is a full save, keep it that way
+    } else if (dst.partial) {
+        delete dst.partial;
+    }
+
+    return dst;
+}
+
 export default DS.Model.extend({
+    pendingDelete: false,
+    pendingSave: null,
+
+    isDeleted: Ember.computed('currentState', 'pendingDelete', function() {
+        return get(this._internalModel.currentState, 'isDeleted') || this.get('pendingDelete');
+    }).readOnly(),
+
     save(options) {
+        if (currentActionModel && currentActionModel.get('isSaving') &&
+                // allow coalesced requests through since we're inside the main save call
+                !(options && options.adapterOptions && options.adapterOptions.coalesce &&
+                  options.adapterOptions.coalesce.type === 'extra')) {
```

Comment:
Can you refactor this boolean into `savingAndNotExtra` or some other descriptive name of the boolean?